packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesPackage.kt

```diff
@@ -55,25 +65,63 @@ class UpdatesPackage : Package {
         }
         // WHEN_VERSIONING_REMOVE_TO_HERE
       }
+    }
+    return listOf(handler)
+  }
 
-      @UiThread
-      private fun shouldAutoSetup(context: Context): Boolean {
-        if (mShouldAutoSetup == null) {
-          mShouldAutoSetup = try {
-            val pm = context.packageManager
-            val ai = pm.getApplicationInfo(context.packageName, PackageManager.GET_META_DATA)
-            ai.metaData.getBoolean("expo.modules.updates.AUTO_SETUP", true)
-          } catch (e: Exception) {
-            Log.e(TAG, "Could not read expo-updates configuration data in AndroidManifest", e)
-            true
+  override fun createReactActivityHandlers(activityContext: Context): List<ReactActivityHandler> {
+    val handler = object : ReactActivityHandler {
+      override fun getDelayLoadAppHandler(activity: ReactActivity, reactNativeHost: ReactNativeHost): ReactActivityHandler.DelayLoadAppHandler? {
+        if (!BuildConfig.EX_UPDATES_ANDROID_DELAY_LOAD_APP) {
+          return null
+        }
+        val context = activity.applicationContext
+        val useDeveloperSupport = reactNativeHost.useDeveloperSupport
+        if (shouldAutoSetup(context) && (useNativeDebug || !useDeveloperSupport)) {
+          return ReactActivityHandler.DelayLoadAppHandler { whenReadyRunnable ->
+            CoroutineScope(Dispatchers.IO).launch {
```

Comment:
I'm not completely familiar with the exception semantics for coroutines, but I assume throwing in the IO thread will cause a hard crash so there's nothing to worry about. Just want to be sure that we always either crash the app or call `whenReadyRunnable`, and never end up in a state where we have a stuck splash screen.

Reply:
i can confirm it will crash the app. tested on kotlin playground and a simple android app.
the playground: https://pl.kotl.in/D087JBswJ